### ISSUE TITLE
Feature Request: Sparse Cholesky decomposition

### ISSUE THREAD

**Issue by rwolst** [tensorflow/tensorflow#15910]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**:
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
- **TensorFlow installed from (source or binary)**:
- **TensorFlow version (use command below)**:
- **Python version**: 
- **Bazel version (if compiling from source)**:
- **GCC/Compiler version (if compiling from source)**:
- **CUDA/cuDNN version**:
- **GPU model and memory**:
- **Exact command to reproduce**:

### Describe the problem
TensorFlow has a Cholesky decomposition [kernel](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/cholesky_op.cc) based on wrappers around the [Eigen](http://eigen.tuxfamily.org/index.php?title=Main_Page) and [cuSOLVER](http://docs.nvidia.com/cuda/cusolver/index.html#cusolver-intro) (for GPUs) libraries.

From experience, a sparse solver can provide huge speedups in the right circumstances. The cuSOLVER library has the feature in their sparse LAPACK library, cuSolverSP, and the Eigen library in the SparseCholesky module.

Alternatively, there is the [CHOLMOD](https://developer.nvidia.com/cholmod) library which is supported by Eigen in the CholmodSupport module. This CHOLMOD library supports both CPU and GPU sparse Cholesky factorisations.

Would a Cholesky decomposition for sparse matrices be a feature of interest?

**Comment by skye** [comment]
@rmlarsen can you comment or redirect? Thanks!

**Comment by dpmerrell** [comment]
Let me revive this issue (and add something to it):

* At present there is a sparse Cholesky factorization hidden deep in TensorFlow. It's a little bit inconvenient, since it requires you to convert your SparseTensor to a "CSRSparseMatrix". It doesn't appear in the documentation.
    - [source code](https://github.com/tensorflow/tensorflow/blob/9664ba19296e58f4437feab4d4b2789cc1e38fd4/tensorflow/core/kernels/sparse/sparse_cholesky_op.cc)
    - [API definition](https://github.com/tensorflow/tensorflow/blob/315ee4a1b9fec30e6e65c025b5b603412baf778a/tensorflow/core/api_def/base_api/api_def_SparseMatrixSparseCholesky.pbtxt)
* I would add another "welcome contribution": a **gradient method** for the sparse Cholesky decomposition. 
    - Backpropagation in this context would maintain the sparse structure. 
    - I.e., if A=LL^T then backprop would only propagate the error in L to the nonzero entries of the lower diagonal of A.

In the source for SparseMatrixSparseCholesky I see a few contributors. @kiszk @penpornk @bloops I wonder if you have thoughts about the sparse Cholesky decomposition and its gradient? 

I have research interests related to this topic and would like to contribute, but I'm finding it difficult to learn TensorFlow's C++ API. I welcome any advice or suggestions.

**Comment by bloops** [comment]
Sure, @dpmerrell, I can help you with the TensorFlow implementation. Do you have the algorithm ready that you want to implement?

**Comment by chunduriv** [comment]
@rwolst,

Please check [SparseMatrixSparseCholesky](https://www.tensorflow.org/jvm/api_docs/java/org/tensorflow/op/linalg/sparse/SparseMatrixSparseCholesky). Thanks!
